Unable to merge junit logs (first worker output empty file)

| Q | A
| --- | ---
| ParaTest version | 7.2.5
| PHPUnit version | 10.3.2
| PHP version | 8.1.22

#### Summary

Same error as was discussed here https://github.com/paratestphp/paratest/issues/745 but with small example.

#### Current behavior

It crashed with exception on joining junit logs files in the end of running tests.

If assert is enabled:
`PHP Fatal error:  Uncaught AssertionError: assert($logFile->isFile() && 0 < (int)$logFile->getSize()) in /var/www/vendor/brianium/paratest/src/JUnit/TestSuite.php:41
Stack trace:
#0 /var/www/vendor/brianium/paratest/src/JUnit/TestSuite.php(41): assert()
#1 /var/www/vendor/brianium/paratest/src/JUnit/LogMerger.php(29): ParaTest\JUnit\TestSuite::fromFile()
#2 /var/www/vendor/brianium/paratest/src/WrapperRunner/WrapperRunner.php(346): ParaTest\JUnit\LogMerger->merge()`

if not ex. occurs later:

`String could not be parsed as XML
...vendor/brianium/paratest/src/JUnit/TestSuite.php:51`


#### How to reproduce: command, code and error stack trace

Reproducible env/error: https://github.com/pistej/parates-bug

#### Expected behavior

I expect junit logs to be merged and report.xml created.

--- a/composer.json
+++ b/composer.json
@@ -43,19 +43,19 @@
         "phpunit/php-code-coverage": "^10.1.3",
         "phpunit/php-file-iterator": "^4.0.2",
         "phpunit/php-timer": "^6.0",
-        "phpunit/phpunit": "^10.3.1",
+        "phpunit/phpunit": "^10.3.2",
         "sebastian/environment": "^6.0.1",
-        "symfony/console": "^6.3.2",
-        "symfony/process": "^6.3.2"
+        "symfony/console": "^6.3.4",
+        "symfony/process": "^6.3.4"
     },
     "require-dev": {
         "ext-pcov": "*",
         "ext-posix": "*",
         "doctrine/coding-standard": "^12.0.0",
         "infection/infection": "^0.27.0",
-        "phpstan/phpstan": "^1.10.26",
-        "phpstan/phpstan-deprecation-rules": "^1.1.3",
-        "phpstan/phpstan-phpunit": "^1.3.13",
+        "phpstan/phpstan": "^1.10.32",
+        "phpstan/phpstan-deprecation-rules": "^1.1.4",
+        "phpstan/phpstan-phpunit": "^1.3.14",
         "phpstan/phpstan-strict-rules": "^1.5.1",
         "squizlabs/php_codesniffer": "^3.7.2",
         "symfony/filesystem": "^6.3.1"

--- a/src/WrapperRunner/WrapperRunner.php
+++ b/src/WrapperRunner/WrapperRunner.php
@@ -237,11 +237,11 @@ final class WrapperRunner implements RunnerInterface
 
     private function destroyWorker(int $token): void
     {
-        // Mutation Testing tells us that the following `unset()` already destroys
-        // the `WrapperWorker`, which destroys the Symfony's `Process`, which
-        // automatically calls `Process::stop` within `Process::__destruct()`.
-        // But we prefer to have an explicit stops.
         $this->workers[$token]->stop();
+        // We need to wait for ApplicationForWrapperWorker::end to end
+        while ($this->workers[$token]->isRunning()) {
+            usleep(self::CYCLE_SLEEP);
+        }
 
         unset($this->workers[$token]);
     }

--- a/test/Unit/WrapperRunner/WrapperRunnerTest.php
+++ b/test/Unit/WrapperRunner/WrapperRunnerTest.php
@@ -611,6 +611,22 @@ EOF;
         self::assertSame(RunnerInterface::SUCCESS_EXIT, $runnerResult->exitCode);
     }
 
+    #[RequiresPhp('8.2')]
+    public function testFunctionalParallelizationWithJunitLogging(): void
+    {
+        $outputFile = $this->tmpDir . DIRECTORY_SEPARATOR . 'test-output.xml';
+
+        $this->bareOptions['path']             = $this->fixture('function_parallelization_tests');
+        $this->bareOptions['--processes']      = '1';
+        $this->bareOptions['--functional']     = true;
+        $this->bareOptions['--max-batch-size'] = 1;
+        $this->bareOptions['--log-junit']      = $outputFile;
+
+        $runnerResult = $this->runRunner();
+        self::assertStringContainsString('.......', $runnerResult->output);
+        self::assertSame(RunnerInterface::SUCCESS_EXIT, $runnerResult->exitCode);
+    }
+
     public function testProcessIsolation(): void
     {
         $this->bareOptions['path']                = $this->fixture('process_isolation' . DIRECTORY_SEPARATOR . 'FooTest.php');

